crates/kitsune_p2p/fetch/src/source.rs

```diff
@@ -0,0 +1,333 @@
+/// not so high that it wastes time for this node trying to talk to a source that is not responding.
+const NUM_PROBE_ATTEMPTS: u32 = 10;
+
+/// A source to fetch from: either a node, or an agent on a node
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum FetchSource {
+    /// An agent on a node
+    Agent(KAgent),
+}
+
+/// Fetch item within the fetch queue state.
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct Sources {
+    inner: IndexSet<FetchSource>,
+    index: usize,
+}
+
+impl Deref for Sources {
+    type Target = IndexSet<FetchSource>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl Sources {
+    pub(crate) fn new(queue: impl IntoIterator<Item = FetchSource>) -> Self {
+        Self {
+            inner: queue.into_iter().collect(),
+            index: 0,
+        }
+    }
+
+    pub(crate) fn next(
+        &mut self,
+        mut state_filter: impl FnMut(&FetchSource) -> bool,
+    ) -> Option<FetchSource> {
+        for _ in 0..self.inner.len() {
+            let fetch_index = self.index;
+            self.index = (self.index + 1) % self.inner.len();
+
+            if let Some(source) = self.inner.get_index(fetch_index) {
+                if state_filter(source) {
+                    return Some(source.clone());
+                }
+            }
+        }
+
+        None
+    }
+
+    pub(crate) fn add(&mut self, source: FetchSource) {
+        self.inner.insert(source);
+    }
+
+    pub(crate) fn retain(&mut self, filter: impl Fn(&FetchSource) -> bool) {
+        self.inner.retain(filter);
+
+        // Ensure the index is still valid
+        if !self.inner.is_empty() {
+            self.index %= self.inner.len();
+        }
+    }
+}
+
+/// The state of a source
+#[derive(Debug, Default)]
+pub(crate) struct SourceState {
+    /// The current state of the source
+    current_state: SourceCurrentState,
+}
+
+impl SourceState {
+    /// Check whether this source should be used. If this source is currently considered available then it will always be usable.
+    /// Otherwise, when the source is in a backoff state, it will only be usable if the backoff is ready. The backoff will be ready
+    /// a fixed number of times to probe the source before going back into a backoff state. If any fetches from the
+    /// probe attempts succeed then the source will be considered available again.
+    pub fn should_use(&mut self) -> bool {
+        match &mut self.current_state {
+            SourceCurrentState::Available(_) => true,
+            SourceCurrentState::Backoff(backoff) => backoff.is_ready(),
+        }
+    }
+
+    /// Check the state of this source. If the source has had too many timeouts then it is still considered valid but it will be put into a backoff state.
+    /// If the source is in a backoff state state and the backoff has expired then the check fails and this source should be dropped.
+    pub fn check(&mut self, config: FetchConfig) -> bool {
```

Comment:
take it or leave it
```suggestion
    pub fn is_valid(&mut self, config: FetchConfig) -> bool {
```

Reply:
Yeah this is a better name although it feels a little strange to use `is_*` with an `&mut` method. I'll go for this rename and consider renaming again if we come up with a better name at some point.

Reply:
I thought so too, ideally this function that returns a `bool` doesn't produce side effects.